Feature/calculate cidr addresses

mike-lang opened the pull request:
Adds the inetAddress property to the return object, which should contain an array of CIDR notation values that fit the ranges specified in our custom notation format.  While testing I discovered that the response sometimes overspecifies the range, returning the occasional redundant cidr address.  For example:

1.2.3.4-10 can be represented as 2 CIDR notation ranges:
- 1.2.3.4/30
- 1.2.3.8/31

but the underlying ip-subnet-calculator module returns three: The mentioned 2, and an additional, redundant
range: 1.2.3.10/32 (which is wholly contained in 1.2.3.8/31)

We can take it as an opportunity for improvement to tighten up the ip-subnet-calculator to return the minimum necessary CIDR notation values.

mike-lang committed:
Add inetAddresses property

Calculates ranges in CIDR notation that fit the specified ranges
in our custom format and puts the array of CIDR notation values into
an array on the inetAddresses property

mike-lang committed:
Added tests on cidr notation conversion

mike-lang committed:
Bump version number

mike-lang commented:
@sberan Can you take a look over my updates to produce equivalent CIDR notation values for ranges?

sberan commented on range.js:
```diff
@@ -16,6 +22,10 @@ exports.range = varArgs(function(specs) {
     var octets = spec.split(".");
 
+    if(octets[0].indexOf('-') != -1 || octets[1].indexOf('-') != -1) {
+      errors.push("Hyphenated values not allowed in first or second octet. ");
```
we should probably disallow `*` here too

mike-lang replied:
Agreed.  Adding

sberan commented on range.js:
```diff
@@ -40,11 +50,34 @@ exports.range = varArgs(function(specs) {
+        fromAddress = prefix + parsedSpec[2][0].toString() + '.0';
+        toAddress = prefix + parsedSpec[2][1].toString() + '.255';
+        masks = ipCalc.calculate(fromAddress, toAddress);
+        Array.prototype.push.apply(inetAddresses, masks.map(mapCalculatorResponseToCIDRNotation));
```
I'm curious why we are using `Array.prototype.push.apply` instead of `inetAddresses.push` here

mike-lang replied:
It's just to make the code a one-liner.  We have arrays in both parameters.  Alternatively, I could forEach on the map result, pushing each value onto inetAddresses.

sberan replied:
no, that's cool. good job.

sberan commented on range.js:
```diff
@@ -40,11 +50,34 @@ exports.range = varArgs(function(specs) {
+  if (errors.length === 0) {
+    parsedSpecs.forEach(function(parsedSpec) {
+      var prefix = parsedSpec[0][0].toString() + '.' + parsedSpec[1][0] + '.';
+       // * in the 4th octet, any variation in the third results in a contiguous block of IPs
```
cool!

sberan commented on test.js:
```diff
@@ -18,3 +20,34 @@ assert(!ip.range('1.0.1.*').contains('1.0.0.10'));
+// Handy tool for generating test data: http://www.ipaddressguide.com/cidr#range
+// If the range finishes with wildcards, we should have
+// one cidr notation range with a netmask.
+assert(ip.range('1.0.1.*').inetAddresses.length === 1);
```
feel free to use `assert.equal` and `assert.notEqual` here for better failure messages ;)

mike-lang replied:
Ha!  Good call

sberan commented on test.js:
```diff
@@ -18,3 +20,34 @@ assert(!ip.range('1.0.1.*').contains('1.0.0.10'));
+// Non-contiguous ranges need multiple cidr notation ranges
+// to specify as well
+assert(ip.range('1.2.0-5.0').inetAddresses.length === 6);
+assert(ip.range('1.2.0-5.0').inetAddresses.indexOf('1.2.0.0/32') !== -1);
```
instead of repeating these calls, we could do with a simple `assert.equal( ip.range('1.2.0-5.0').inetAddresses, ['1.2.0.0/32', '1.2.1.0/32', ....]` here eh?

sberan replied:
sorry, `assert.deepEqual`

mike-lang replied:
I couldn't remember what order the module puts the results in :)

sberan replied:
`.sort()` them first and look at the failure message?

mike-lang replied:
I like it.  Incoming commit

sberan commented:
a few comments, otherwise LGTM!

mike-lang committed:
Updates in response to review

Disallow wildcards in first or second octets, use assert.equals &
assert.notEquals

Added a few more comments

mike-lang commented:
Good deal.  I made edits along the suggested lines.  I'll merge so I can pull this into my cms project locally
